Parallel assignment

rrosenblum opened the pull request:
This fixes #1136.

I have been working on this cop for quite some time, and I think that I finally have all the kinks worked out. It will only register an offense when the number of variables being assigned to matches the number of assigning variables. It will also allow you to perform variable reassignment (`a, b = b, a`). Any assignment using splat or a method is also allowed.

I think that this will be able to detect a large number of offenses while allowing parallel assignment where you would want to use it. 

Please look over it and let me know if you have any concerns with this cop.

rrosenblum commented on lib/rubocop/cop/style/space_around_block_parameters.rb:
```diff
@@ -14,17 +14,26 @@ module Style
+        EXTRA_BEFORE = 'Extra space before'
+        EXTRA_BEFORE_FIRST = 'Extra space before first'
+        EXTRA_AFTER = 'Extra space after last'
+        AFTER_CLOSING_PIPE = 'after closing `|`'
```
Splitting mass assignment into separate assignments caused a method too long error on `check_inside_pipes`. I extracted the strings into constants so that I could fit some of the statements on one line and shorten the length of the method. Alternatively, I could extract each condition into their own method. 

I can extract this change out into a different commit if need be.

rrosenblum commented on lib/rubocop/cop/util.rb:
```diff
@@ -14,10 +14,10 @@ module Util
+
+      OPERATOR_METHODS = [
+        :|, :^, :&, :<=>, :==, :===, :=~, :>, :>=, :<, :<=, :<<, :>>, :+,
+        :-, :*, :/, :%, :**, :~, :+@, :-@, :[], :[]=, :!, :!=, :!~, :`]
```
This was a similar change as before. This file was failing for module length too long. This seemed like an easy place to save a line.

rrosenblum commented on lib/rubocop/cop/util.rb:
```diff
@@ -184,12 +185,12 @@ def directions(side)
       def begins_its_line?(range)
-        source_before_end = range.source_buffer.source[0...range.begin_pos]
-        source_before_end =~ /\n\s*\Z/
+        range.source_buffer.source[0...range.begin_pos] =~ /\n\s*\Z/
```
This was another line that had to change to shorten the module.

rrosenblum commented on spec/rubocop/cop/performance/parallel_assignment_spec.rb:
```diff
@@ -0,0 +1,147 @@
+  end
+
+  it_behaves_like('allowed', 'a = 1')
+  it_behaves_like('allowed', "a = *'foo'")
```
`*'foo'` is actually valid syntax. It will evaluate out to `["foo"]`.

rrosenblum commented:
I had commented on this code earlier, but when I rebased it, the comments went away.

rrosenblum commented:
I have split the fixing of offenses into 2 commits.

bbatsov commented:
Btw, see https://github.com/bbatsov/rubocop/pull/1833 as it makes some changes to the auto-correct API.

rrosenblum commented:
I have changed the number of allowed offenses for method length and module length.

jonas054 commented on lib/rubocop/cop/performance/parallel_assignment.rb:
```diff
@@ -0,0 +1,80 @@
+
+          return if left_elements.any? do |le|
+            right_elements.any? do |re|
+              re.respond_to?('loc') &&
```
In what situation doesn't this object respond to `loc`? It's not covered by the specs.

rrosenblum replied:
When I had the code written to detect errors when the left variables did not match the right variables, this was necessary because methods return symbols instead of nodes. It appears that this can be removed now.

jonas054 commented on lib/rubocop/cop/performance/parallel_assignment.rb:
```diff
@@ -0,0 +1,80 @@
+          # allow mass assignment when using splat
+          return if (left_elements + right_elements).compact.any?(&:splat_type?)
+
+          return if left_elements.any? do |le|
```
Extract this little algorithm to its own method, perhaps called `swap_assignment?`. Then the purpose becomes more clear.

jonas054 commented on lib/rubocop/cop/performance/parallel_assignment.rb:
```diff
@@ -0,0 +1,80 @@
+            groups = l_vars.zip(r_vars)
+
+            assignment = groups.map do |pair|
+              pair[1] = 'nil' if pair[1].nil?
```
This line is not covered by specs, i.e., it can be removed and everything still passes.

rrosenblum replied:
Sorry about that, I originally had coded this up to work with mismatched parameters on the left and right, but I decided that the detection/auto correction was risky with that configuration. I must have missed this in the cleanup.

jonas054 commented on lib/rubocop/cop/performance/parallel_assignment.rb:
```diff
@@ -0,0 +1,80 @@
+
+        def extract_sources(node)
+          if node.children.size == 1
+            [node.loc.expression.source]
```
This branch is not covered by specs.

rrosenblum commented:
@jonas054, sorry about those bits of code that were not covered by tests. I originally was trying to make this cop be able to account for situations where the number of variables on the left and right side did not match. This was proving troublesome for detection and correction so I changed the cop to only complain when the variable count matches. I missed removing a couple of lines of code. I have removed those lines of codes, and I have made some small modifications to the code to account for edge cases that I had not thought of before.

jonas054 commented on lib/rubocop/cop/performance/parallel_assignment.rb:
```diff
@@ -0,0 +1,81 @@
+            r_vars = extract_sources(right)
+            groups = l_vars.zip(r_vars)
+
+            assignment = groups.map do |pair|
```
The block fits on a single line now.

jonas054 commented:
Just one remark on the code, and the commit message should be "Increase ..." rather than "Increasing ...". Other than that, it looks fine now.

rrosenblum commented:
The code has been updated.

bbatsov commented:
:+1:

sferik commented:
:+1: Happy to see this in RuboCop 0.31.0. In addition to making the code more readable, it also [makes it somewhat faster](https://github.com/JuanitoFatas/fast-ruby#parallel-assignment-vs-sequential-assignment-code).

rrosenblum commented:
@sferik, thanks for the positive feedback.
:+1:
